### PR TITLE
Add new collections, optimize, refactor and fix existing code

### PR DESCRIPTION
**Pull request by spisladqo**
Add `vertices` array where index is the vertex id.

Add `outEdgesMap`, that for every vertex (key) contains an array of edges, coming from this vertex (data):
- in a directed graph any edge from such array will always have first vertex similar to the key.
- in an undirected graph it's not obligatory.

Optimize code using new collections.

Delete repeating code using `super` invocations.

Rename some variables and functions for better understanding of their purpose.

### CODE REVIEW

File: app/src/main/kotlin/model/abstractGraph/Graph.kt [hunk 1]
```diff
@@ -1,34 +1,37 @@
...
 abstract class Graph<D> {
-    protected val adjacencyMap: MutableMap<Vertex<D>, ArrayList<Vertex<D>>> = mutableMapOf()
+    protected val vertices: ArrayList<Vertex<D>> = arrayListOf()
     protected val edges: MutableSet<Edge<D>> = mutableSetOf()
-    private var currentId = 0
+
+    protected val adjacencyMap: MutableMap<Vertex<D>, ArrayList<Vertex<D>>> = mutableMapOf()
+    protected val outEdgesMap: MutableMap<Vertex<D>, ArrayList<Edge<D>>> = mutableMapOf()
+
+    private var nextId = 0
```
Comment (qrutyy) [inline app/src/main/kotlin/model/abstractGraph/Graph.kt:10]: 😃

File: app/src/main/kotlin/model/abstractGraph/Graph.kt [hunk 2]
```diff
@@ -1,34 +1,37 @@
 package model.abstractGraph
 
 abstract class Graph<D> {
-    protected val adjacencyMap: MutableMap<Vertex<D>, ArrayList<Vertex<D>>> = mutableMapOf()
+    protected val vertices: ArrayList<Vertex<D>> = arrayListOf()
     protected val edges: MutableSet<Edge<D>> = mutableSetOf()
-    private var currentId = 0
+
+    protected val adjacencyMap: MutableMap<Vertex<D>, ArrayList<Vertex<D>>> = mutableMapOf()
+    protected val outEdgesMap: MutableMap<Vertex<D>, ArrayList<Edge<D>>> = mutableMapOf()
```
Comment (shakareem) [inline app/src/main/kotlin/model/abstractGraph/Graph.kt]: mb rename to outgoingEdgesMap?

File: app/src/main/kotlin/model/abstractGraph/Graph.kt [hunk 3]
```diff
@@ -1,34 +1,37 @@
...
-        adjacencyMap.remove(vertexToRemove)
-
         return vertexToRemove
     }
 
-    abstract fun addEdge(vertex1: Vertex<D>, vertex2: Vertex<D>): Edge<D>
-
-    abstract fun removeEdge(edgeToRemove: Edge<D>): Edge<D>
-
+    // TODO: RENAME PARAMETER NAME
```
Comment (shakareem) [inline app/src/main/kotlin/model/abstractGraph/Graph.kt]: Remove TODO.

File: app/src/main/kotlin/model/DirectedGraph.kt [hunk 4]
```diff
@@ -9,11 +9,13 @@ open class DirectedGraph<D> : Graph<D>() {
         if (vertex1 == vertex2)
             throw IllegalArgumentException("Can't add edge from vertex to itself.")
 
-        if (vertex1 !in adjacencyMap.keys || vertex2 !in adjacencyMap.keys)
-            throw NoSuchElementException("Vertex1 or vertex2 is not in the adjacency map.")
+        if (vertex1.id > vertices.size || vertex2.id > vertices.size)
+            throw NoSuchElementException("Vertex1 or vertex2 is not in the vertices array.")
```
Comment (shakareem) [inline app/src/main/kotlin/model/DirectedGraph.kt:13]: Nice!!

File: app/src/main/kotlin/model/abstractGraph/Graph.kt [hunk 5]
```diff
@@ -42,24 +45,29 @@ abstract class Graph<D> {
             }
         }
 
-        removeVertexFromIncidentEdgesAndAdjacentVerticesMapValues(lastAddedVertex)
-        adjacencyMap.remove(lastAddedVertex)
+        removeVertexFromEverywhere(lastAddedVertex)
     }
 
-    private fun removeVertexFromIncidentEdgesAndAdjacentVerticesMapValues(vertexToRemove: Vertex<D>) {
+    private fun removeVertexFromEverywhere(vertexToRemove: Vertex<D>) {
```
Comment (qrutyy) [inline app/src/main/kotlin/model/abstractGraph/Graph.kt:54]: We need to add vertex deletion from the 'vertices' array. 😃

**Review by qrutyy — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/model/abstractGraph/Graph.kt (hunk 5 above)
Comment (shakareem) [inline app/src/main/kotlin/model/abstractGraph/Graph.kt:54]: There is a bug which i will fix! I promise!

**Review by shakareem — CHANGES_REQUESTED** [review]
I have to fix a bug, you have to rename variables and remove comment.

File: app/src/main/kotlin/model/WeightedDirectedGraph.kt [hunk 6]
```diff
@@ -6,20 +6,12 @@ import java.util.*
 import kotlin.NoSuchElementException
 
 class WeightedDirectedGraph<D> : DirectedGraph<D>() {
-
     val weightMap: MutableMap<Edge<D>, Int> = mutableMapOf()
```
Comment (kar1mgh) [inline app/src/main/kotlin/model/WeightedDirectedGraph.kt:9]: We should make it private set.

Comment (shakareem) [inline app/src/main/kotlin/model/WeightedDirectedGraph.kt]: We should make it private set.

**Review by shakareem — CHANGES_REQUESTED** [review]
We should make weightmap private set.

File: app/src/main/kotlin/model/abstractGraph/Graph.kt (hunk 1 above)
Comment (spisladqo) [reply app/src/main/kotlin/model/abstractGraph/Graph.kt:10]: I do not agree.

**Review by qrutyy — APPROVED** [review]
I like it)
Good job, guys!

**Review by shakareem — APPROVED** [review]

**Pull request merged**